src/bijectors/coupling_layer.jl

```diff
@@ -0,0 +1,244 @@
+using SparseArrays
+
+"""
+    PartitionMask{A}(A_1::A, A_2::A, A_3::A) where {A}
+
+This is used to partition and recombine a vector into 3 disjoint "subvectors".
+
+Implements
+- `partition(m::PartitionMask, x)`: partitions `x` into 3 disjoint "subvectors"
+- `combine(m::PartitionMask, x_1, x_2, x_3)`: combines 3 disjoint vectors into a single one
+
+Note that `PartitionMask` is _not_ a `Bijector`. It is indeed a bijection, but
+does not follow the `Bijector` interface.
+
+Its main use is in `CouplingLayer` where we want to partition the input into 3 parts,
+one part to transform, one part to map into the parameter-space of the transform applied
+to the first part, and the last part of the vector is not used for anything.
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= assumes input-length 3
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0,
+  [2, 1]  =  1.0,
+  [3, 1]  =  1.0)
+
+julia> # Partition into 3 parts; the last part is inferred to be indices `[3, ]` from
+       # the fact that `[1]` and `[2]` does not make up all indices in `1:3`.
+       x1, x2, x3 = partition(m, [1., 2., 3.])
+([1.0], [2.0], [3.0])
+
+julia> # Recombines the partitions into a vector
+       combine(m, x1, x2, x3)
+3-element Array{Float64,1}:
+ 1.0
+ 2.0
+ 3.0
+```
+
+"""
+struct PartitionMask{A}
+    A_1::A
+    A_2::A
+    A_3::A
+
+    # Only make it possible to construct using matrices
+    PartitionMask(A_1::A, A_2::A, A_3::A) where {A <: AbstractMatrix{<:Real}} = new{A}(A_1, A_2, A_3)
+end
+
+function PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::AbstractVector{Int}
+)
+    A_1 = spzeros(n, length(indices_1));
```

Comment:
```suggestion
    A_1 = spzeros(Bool, n, length(indices_1))
```

Reply:
Similar to others

Reply:
Is this actually more efficient when it comes to matrix multiplication and such? Thinking about both CPU and GPU perf here.

Reply:
Alternatively, for convenience we can do
```julia
function PartitionMask(
    n::Int,
    indices_1::AbstractVector{Int},
    indices_2::AbstractVector{Int},
    indices_3::AbstractVector{Int},
   T::Type = Float64
)
    A_1 = spzeros(T, n, length(indices_1));
```

Reply:
I'm not sure about the perfomrance. This is attempting to solve the type issue.

Your suggestion works only if users manually input the type. I don't think it's ideal to have this. Why partitioning cannot preserve the array type it works on?